keps/sig-instrumentation/4785-resource-state-metrics/README.md

```diff
@@ -0,0 +1,1216 @@
+Resource State API, while maintaining a 3x faster round trip time for metric
+generation.
+
+- At its core, the controller relies on its managed resource,
+`ResourceMetricsMonitor` to fetch the metric generation configuration. Parts of the
+configuration may be defined using different `resolver`s, such as `unstructured`
+or `CEL`.
+- Once fetched, the controller `unmarshal`s the configuration YAML directly into
+`stores` which are a set of metric `families`, which in turn are a set of 
+`metrics`.
+- Metric `stores` are created based on its respective GVKR (a type that embeds 
+`schema.GroupVersionKind`, `schema.GroupVersionResource` to avoid 
+[plural ambiguities]), and reflectors for the specified resource are
+initialized, and populate the stores on its update.
+- `/metrics` pings on `RSM_MAIN_PORT` trigger the server to write the
+raw metrics, combined with its appropriate header(s), in the response. All
+generated metrics are hardcoded to `gauge`s by design, as Prometheus lacks
+support for some OpenMetrics-specified metrics' types, such as `Info` and
+`StateSets`.
+
+At the moment, the `spec` houses a single `configuration` field, which defines
+the metric generation configuration as follows (please note that the schema is
+fast-moving at this point and may be subject to change based on the [feedback 
+obtained](https://github.com/rexagod/resource-state-metrics/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen)):
+
+```yaml
+stores: # Set of metrics stores for each CR we want to generate metrics for.
+  - g: "contoso.com" # CR's group.
+    v: "v1alpha1"    # CR's version.
+    # Both kind and resource names are required to avoid plural ambiguities, see
+    # https://github.com/kubernetes-sigs/kubebuilder/issues/3402.
+    k: "MyPlatform"  # CR's kind.
+    r: "myplatforms" # CR's resource.
+    selectors: # Set of filters to narrow down the selected CRs, may be:
+      field: "metadata.namespace=default" # field selector(s), and (/or),
+      label: "app.kubernetes.io/part-of=sample-controller" # label selector(s).
+    families: # Set of metrics families to generate for the specified CR.
+      - name: "platform_info" # The metric family name, plugged in as-is.
+        help: "Information about a MyPlatform instance" # The help text for the
+                                                        # metric family, plugged
+                                                        # in as-is.
+        metrics: # Set of metrics to generate under the current metrics family.
+          - resolver: "cel" # Preferred resolver to parse the value expressions.
+                            # MAY BE SPECIFIED AT ANY LEVEL.
+                            # INNER RESOLVERS OVERRIDE THE OUTER ONES.
+                            # NO RESOLVER DEFAULTS TO THE "UNSTRUCTURED" ONE.
+            # Set of label-sets to generate for the current metric.
+            labelKeys: # Set of ordered label-keys, static in nature.
+              - "name"
```

Comment:
I suppose that for the vast majority of cases, you'd want a `name` label (as well as a `namespace`label for namespaced custom resources). Should it be enabled by default?

Reply:
+1 towards the idea of having a way to specify default labels. I don't know if it should necessarily be enabled by default and fixed to `name` and `namespace` only. Maybe we add a new field at the same level as the g,v,k,r in the config called defaultLabels and allow users to specify the one they want to be applied to all the metrics that will follow. The result would be kind of similar to what we have today in ksm: https://github.com/kubernetes/kube-state-metrics/blob/0b01e3abce1da521b5e620b8aaa76774bb0fda87/internal/store/pod.go#L39

Reply:
Default labels may be defined at any scope, [like so](https://github.com/rexagod/resource-state-metrics/blob/main/manifests/custom-resource.yaml#L93-L103). These are not enabled by default to allow creating scalar metrics if users want to.